parser-typechecker/src/Unison/Syntax/TermParser.hs

```diff
@@ -161,27 +161,33 @@ matchCases1 start = do
 --   (42, x) -> ...
 matchCase :: Var v => P v (Int, [Term.MatchCase Ann (Term v Ann)])
 matchCase = do
-  pats <- sepBy1 (reserved ",") parsePattern
+  pats <- sepBy1 (label "\",\"" $ reserved ",") parsePattern
   let boundVars' = [v | (_, vs) <- pats, (_ann, v) <- vs]
       pat = case fst <$> pats of
         [p] -> p
         pats -> foldr pair (unit (ann . last $ pats)) pats
       unit ann = Pattern.Constructor ann (ConstructorReference DD.unitRef 0) []
       pair p1 p2 = Pattern.Constructor (ann p1 <> ann p2) (ConstructorReference DD.pairRef 0) [p1, p2]
-  guardsAndBlocks <- many $ do
-    guard <-
-      asum
-        [ Nothing <$ P.try (reserved "|" *> quasikeyword "otherwise"),
-          optional $ reserved "|" *> infixAppOrBooleanOp
-        ]
-    t <- block "->"
-    pure (guard, t)
+  let guardedBlocks = label "pattern guard" . some $ do
```

Comment:
We now parse guarded and unguarded blocks separately, each pattern either has one or more guards, or no guards at all, you can't mix and match guarded and unguarded RHSs.